firestore/functions/src/index.ts

```diff
@@ -0,0 +1,111 @@
+import * as functions from 'firebase-functions';
+import * as admin from 'firebase-admin';
+import * as util from 'util';
+import { DocumentSnapshot, DocumentData } from '@google-cloud/firestore';
+// // Start writing Firebase Functions
+// // https://firebase.google.com/docs/functions/typescript
+//
+
+"use strict";
+
+// Uncomment these lines to run our Cloud Functions in the emulator
+
+// const sa = require( './super_secret/new-test-projects-a746c-firebase-adminsdk-ttdbm-3227776c2b.json');
+ 
+// admin.initializeApp({
+//   credential: admin.credential.cert(sa),
+//   databaseURL: "http://localhost:8080/"
+// });
+ 
+admin.initializeApp();
+
+/**
+ * Demonstrate a basic working cloud function
+ */
+export const helloWorld = functions.https.onRequest((request, response) => {
+ response.send("Hello from Firebase!");
+});
+
+/**
+ * Callable cloud function to get back a list of favorite restaurants for a user
+ */
+export const getFavorites_v0 = functions.https.onCall(async (data, context) => {
+    const userID = data.uid;
+    try {
+        if (userID) {
+            console.log(`I am going to get favorites for user ${userID}`);
+            const userDoc = await admin.firestore().doc(`users/${userID}`).get();
+            console.log(`Here is my user doc`);
+            console.log(util.inspect(userDoc.data()));
+            const userData = userDoc.data();
+            if (!userData) {
+                console.error('Data not found for ', userID);
+                return([]);
+            }
+            const favorites = userData.favorites;
+            if (!favorites) {
+                console.log('User has no favorites');
+                return([]);
+            }
+            const fetchPromises: Promise<DocumentSnapshot>[] = [];
+            favorites.forEach((restID: string)  => {
+                console.log('I\'m gonna fetch data for ', restID);
+                const nextPromise = admin.firestore().doc(`restaurants/${restID}`).get();
+                fetchPromises.push(nextPromise);
+            });
+            return Promise.all(fetchPromises).then((docSnaps) => {            
+                const responseArray: DocumentData[] = []
+                docSnaps.forEach((snapshot: DocumentSnapshot) => {
+                    const docData = snapshot.data()!;
+                    docData['.id'] = snapshot.id;
+                    responseArray.push(docData);
+                });
+                return(responseArray);
+            });
```

Comment:
Since you're using `async/await` syntax higher up, you can do that here too:

```ts
const docSnaps = await Promise.all(fetchPromises);
const responseArray: DocumentData[] = [];

docSnaps.forEach((snapshot: DocumentSnapshot) => {
  const docData = snapshot.data()!;
  docData[".id"] = snapshot.id;
  responseArray.push(docData);
});

return responseArray;
```

if you want to get really fancy:

```ts
const docSnaps = await Promise.all(fetchPromises);
return docSnaps.map((snapshot: DocumentSnapshot) => {
  return {
    ".id": snapshot.id,
    ...snapshot.data()
  };
});
```

Reply:
Oh, good call. I added the await, and will note during the video that this could be made fancier with a map and spread.

Reply:
Updated to use async/await. While I did end up using `map`, I think `spread` might be a little too fancy for our target audience. I'll call it out, though,